File: Data Structures/Graph-based-DSAs/dfs.py
"""DFS implementation in python: """

def dfs(graph, start, visited=None):
    if visited is None:
        visited = set()
    visited.add(start)

    print(start)

    for next in [x for x in graph[start] if x not in visited]:
        dfs(graph, next, visited)
    return visited

if __name__ == "__main__":
    graph = {'0': ['1', '2', '3'],
         '1': ['0', '2'],
         '2': ['0', '1', '4'],
         '3': ['0'],
         '4': ['2']}

    print(dfs(graph, '0'))

File: Data Structures/Graph-based-DSAs/dfs.cpp
#include "iostream"
#include "bits/stdc++.h"

using namespace std;

class Graph{
    public:
        Graph(int V);
        void add_edge(int src, int dest);
        void DFS(int vertex);
    private:
        int numVertices;
        list<int> *adjList;
        bool *visited;
};

Graph::Graph(int V){
    numVertices = V;
    adjList = new list <int> [numVertices];
    visited = new bool[numVertices];   
}

void Graph::add_edge(int src, int dest){
    adjList[src].push_front(dest);
}

void Graph::DFS(int vertex){
    visited[vertex] = true;
    cout << vertex << " ";
    list<int>::iterator it = adjList[vertex].begin();
    for (it; it != adjList[vertex].end(); ++it){
        if (!visited[*it]){
            DFS(*it);
        }
    }

}

int main(){
    Graph g(5);
    g.add_edge(0, 1);
    g.add_edge(0, 2);
    g.add_edge(0, 3);
    g.add_edge(1, 2);
    g.add_edge(2,4);

    g.DFS(0);
}